File: leaf-system/src/main/java/com/leaf/system/common/interceptor/AuditInterceptor.java
package com.leaf.system.common.interceptor;

import com.leaf.system.domain.BaseEntity;
import com.leaf.system.domain.BaseEntityDraft;
import com.leaf.system.domain.BaseEntityProps;
import jakarta.enterprise.context.ApplicationScoped;
import org.babyfish.jimmer.ImmutableObjects;
import org.babyfish.jimmer.sql.DraftInterceptor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;

/**
 * Jimmer 拦截器完成审计功能
 *
 * @author jhlz
 * @version 1.0.0
 */
@ApplicationScoped
public class AuditInterceptor implements DraftInterceptor<BaseEntity, BaseEntityDraft> {
    @Override
    public void beforeSave(@NotNull BaseEntityDraft draft, @Nullable BaseEntity baseEntity) {
        if (!ImmutableObjects.isLoaded(draft, BaseEntityProps.UPDATE_TIME)) {
            draft.setUpdateTime(LocalDateTime.now());
        }

    }
}

File: src/main/kotlin/com/jhlz/common/interceptor/AuditInterceptor.kt
package com.jhlz.common.interceptor

import com.jhlz.domain.BaseEntity
import com.jhlz.domain.BaseEntityDraft
import io.micronaut.context.annotation.Factory
import org.babyfish.jimmer.kt.isLoaded
import org.babyfish.jimmer.sql.DraftInterceptor
import java.time.OffsetDateTime

/**
 * Jimmer 拦截器完成审计功能
 *
 * DraftPreProcessor 与 DraftInterceptor 关键不同就是无法区分是否是插入还是删除操作
 * 可以说是 DraftInterceptor 的弱化版本，但是性能好，少了一次额外的查询
 * @author jhlz
 * @version 1.0.0
 */
@Factory
class AuditInterceptor : DraftInterceptor<BaseEntity, BaseEntityDraft> {

    override fun beforeSave(draft: BaseEntityDraft, original: BaseEntity?) {
        // 更新
        if (!isLoaded(draft, BaseEntity::updateTime)) {
            draft.updateTime = OffsetDateTime.now()
        }

        // if (!isLoaded(draft, BaseEntity::updator)) {
        //     draft.updator {
        //         id = 1
        //     }
        // }

        // 新增
        if (original === null) {
            if (!isLoaded(draft, BaseEntity::createTime)) {
                draft.createTime = OffsetDateTime.now()
            }

            // if (!isLoaded(draft, BaseEntity::creator)) {
            //     draft.creator() {
            //         id = 1
            //     }
            // }
        }
    }
}
